### PR TITLE
feat(error): Propagate specific errors

### PR DESCRIPTION
**Pull request by kptdobe**
WIP PR for #378 

I have created a deployment with all error cases: 

- [https://alex.helix-demo.xyz/e.403.html](https://alex.helix-demo.xyz/e.403.html)
- [https://alex.helix-demo.xyz/e.404.html](https://alex.helix-demo.xyz/e.404.html)
- [https://alex.helix-demo.xyz/e.500.html](https://alex.helix-demo.xyz/e.500.html)
- [https://alex.helix-demo.xyz/e.502.html](https://alex.helix-demo.xyz/e.502.html)
- [https://alex.helix-demo.xyz/e.503.html](https://alex.helix-demo.xyz/e.503.html)
- [https://alex.helix-demo.xyz/e.504.html](https://alex.helix-demo.xyz/e.504.html)

Error code project source is here: https://github.com/kptdobe/helix-errors

I also merged https://github.com/adobe/helix-publish/pull/379 in this PR to have the full picture.

But I am bit stuck: non of those uses the custom error pages. The 504 for example reveal the `Error 957 Gateway Timeout` internal error message. I do not get yet the flow and what needs to happen so that the `hlx_error_errors` sub routine gets called.
Maybe @trieloff or @stefan-guggisberg you have feedback.

### CODE REVIEW

**Comment by kptdobe** [comment]
Interestingly, the 500 works as expected...

**Comment by kptdobe** [comment]
OMG.. the number mapping + different place handling + different error page depending on error number makes it... is sooo complicated!!
But with extra debugs, I think I am going to make it.

**Comment by stefan-guggisberg** [comment]
> But I am bit stuck: non of those uses the custom error pages. The 504 for example reveal the Error 957 Gateway Timeout internal error message. I do not get yet the flow and what needs to happen so that the hlx_error_errors sub routine gets called.

I have a hard time understanding the flow as well :/ There are 4 different helix error handling subroutines:

* `hlx_fetch_errors` & `hlx_fetch_error` (called from `vcl_fetch`)
* `hlx_error_errors`(called from `vcl_error`)
* `hlx_deliver_errors` (called indirectly from `vcl_deliver`)

The naming refers to the states/phases in the request processing cycle: `fetch`, `error`, `deliver`

I believe the flow is as follows: 

1. `fetch`: interpret origin response and translate to/trigger custom error (e.g. 951)
2. `error`: mapping of error status to synthetic error responses (?)
3. `deliver`: mapping of custom errors to http error status (e,g, 951 -> 404)

@trieloff clarifications & comments are welcome :-)

**Comment by trieloff** [comment]
`hlx_error_errors` selects the error pages, and their content type

`hlx_deliver_errors` sets the HTTP status code and status text

`hlx_fetch_errors` tries to interpret OpenWhisk's error messages (which are all 404s) and turn them into something more specific. This is not relevant for the errors we are trying to propagate here.

`hlx_fetch_error` is called from `hlx_fetch`, i.e. after the backend response has been received (or not).

**Comment by stefan-guggisberg** [comment]
> `hlx_fetch_error` is called from `hlx_fetch`, i.e. after the backend response has been received (or not).

`hlx_fetch_error` is called from ~~`hlx_fetch`~~`vcl_fetch`

File: layouts/fastly/helix.vcl [hunk 1]
```diff
@@ -1231,6 +1231,11 @@ sub hlx_deliver_errors {
 sub hlx_error_errors {
   set obj.http.X-Trace = req.http.X-Trace + "; hlx_error_errors(" obj.status ")";
   # Cache Condition: OpenWhisk Error Prio: 10
+  if (obj.status == 404 ) {
```
Comment (trieloff) [inline layouts/fastly/helix.vcl]: We need to be careful here so that we do not override a custom `404.html` if it has been provided by `helix-dispatch`.

Comment (kptdobe) [reply layouts/fastly/helix.vcl]: for now I am trying to understand why I do not get the default 404.html defined in the VCL.
how should I set up the custom error page ? `/htdocs/404.html` ?

**Comment by kptdobe** [comment]
The biggest pain point to debug the VCL is that X-Trace does not contain the full trace in many cases.

File: layouts/fastly/helix.vcl (hunk 1 above)
Comment (trieloff) [reply layouts/fastly/helix.vcl]: Either `htdocs/404.html` or `/404.html`.

**Comment by kptdobe** [comment]
> I wouldn't mix this with the issue at hand (propagating errors). I'd rather address it with a separate issue.

To debug and move forward, I have to enable some more logs and fix the trace.
But yes, at the end, I might create a separate PR for the trace.

Note this is a WIP PR, still my playground.

File: layouts/fastly/helix.vcl [hunk 2]
```diff
@@ -1206,11 +1212,30 @@ sub hlx_deliver_errors {
...
+
+  if (resp.status == 957) {
+     set resp.status = 504;
+     set resp.response = "Gateway Timeout";
+  }
 }
 
 sub hlx_error_errors {
-  set req.http.X-Trace = req.http.X-Trace + "; hlx_error_errors";
+  set obj.http.X-Trace = req.http.X-Trace + "; hlx_error_errors(" obj.status ")";
```
[the remark this answers is not in the document]
Comment (kptdobe) [reply layouts/fastly/helix.vcl]: yes, this fixed the missing `hlx_error_errors` - well, in some cases... not all. Maybe this is why it is missing now.

File: layouts/fastly/helix.vcl [hunk 3]
```diff
@@ -1475,9 +1515,9 @@ sub vcl_deliver {
   }
 
   # remove temporary headers used to reconstruct trace information
-  unset resp.http.X-PreFetch-Miss;
-  unset resp.http.X-PreFetch-Pass;
-  unset resp.http.X-PostFetch;
+  # unset resp.http.X-PreFetch-Miss;
```
[the remark this answers is not in the document]
Comment (kptdobe) [reply layouts/fastly/helix.vcl]: debugging the trace...

**Comment by kptdobe** [comment]
The current status is: 

- [x] [https://alex.helix-demo.xyz/e.403.html](https://alex.helix-demo.xyz/e.403.html) -> cannot work - see https://github.com/adobe/helix-dispatch/issues/218
- [x] use [https://alex.helix-demo.xyz/e.403.html](https://alex.helix-demo.xyz/.bla.html) instead -> works
- [ ] [https://alex.helix-demo.xyz/e.404.html](https://alex.helix-demo.xyz/e.404.html) -> still need to understand what happens
- [x] [https://alex.helix-demo.xyz/e.500.html](https://alex.helix-demo.xyz/e.500.html) -> works
- [x] [https://alex.helix-demo.xyz/e.502.html](https://alex.helix-demo.xyz/e.502.html) -> works
- [x] [https://alex.helix-demo.xyz/e.503.html](https://alex.helix-demo.xyz/e.503.html) -> works
- [x] [https://alex.helix-demo.xyz/e.504.html](https://alex.helix-demo.xyz/e.504.html) -> works

**Comment by kptdobe** [comment]
The final status is: 

- [x] [https://alex.helix-demo.xyz/e.403.html](https://alex.helix-demo.xyz/e.403.html) -> cannot work - see https://github.com/adobe/helix-dispatch/issues/218
- [x] use [https://alex.helix-demo.xyz/.bla.html](https://alex.helix-demo.xyz/.bla.html) instead -> works
- [x ] [https://alex.helix-demo.xyz/e.404.html](https://alex.helix-demo.xyz/e.404.html) -> works but change needs to be reviewed
- [x] [https://alex.helix-demo.xyz/e.500.html](https://alex.helix-demo.xyz/e.500.html) -> works
- [x] [https://alex.helix-demo.xyz/e.502.html](https://alex.helix-demo.xyz/e.502.html) -> works
- [x] [https://alex.helix-demo.xyz/e.503.html](https://alex.helix-demo.xyz/e.503.html) -> works
- [x] [https://alex.helix-demo.xyz/e.504.html](https://alex.helix-demo.xyz/e.504.html) -> works

I clean up the PR and do "real one".

File: layouts/fastly/helix.vcl [hunk 4]
```diff
@@ -801,7 +801,6 @@ sub hlx_fetch_static {
     set req.http.X-Trace = req.http.X-Trace + "(400)";
     # Cache for a short time, restart will get rid of it anyway
     set beresp.ttl = 60s;
-    return(deliver);
```
Comment (kptdobe) [inline layouts/fastly/helix.vcl]: @trieloff That's the line that causes 404 to not use the default error page.
I tested cases with (404.html at root of the project) and without the custom error page, it renders the expect one.
I just wonder if this does not break something else.

**Comment by kptdobe** [comment]
The final status is: 

- [x] [https://alex.helix-demo.xyz/e.403.html](https://alex.helix-demo.xyz/e.403.html) -> 404 - cannot work - see https://github.com/adobe/helix-dispatch/issues/218
- [x] use [https://alex.helix-demo.xyz/.bla.html](https://alex.helix-demo.xyz/.bla.html) instead -> 403
- [x] [https://alex.helix-demo.xyz/e.404.html](https://alex.helix-demo.xyz/e.404.html) -> 404
- [x] [https://alex.helix-demo.xyz/e.500.html](https://alex.helix-demo.xyz/e.500.html) -> 500
- [ ] [https://alex.helix-demo.xyz/e.502.html](https://alex.helix-demo.xyz/e.502.html) -> 502
- [x] [https://alex.helix-demo.xyz/e.503.html](https://alex.helix-demo.xyz/e.503.html) -> 503
- [x] [https://alex.helix-demo.xyz/e.504.html](https://alex.helix-demo.xyz/e.504.html) -> 504

Ready for review.

File: layouts/fastly/helix.vcl (hunk 3 above)
[the remark this answers is not in the document]
Comment (kptdobe) [reply layouts/fastly/helix.vcl]: removed.

File: layouts/fastly/helix.vcl (hunk 4 above)
Comment (trieloff) [reply layouts/fastly/helix.vcl]: Even if it does, the scope of breaking something else should be quite small.

**Pull request merged**